cpp/inline/mcts/SearchLog.inl

```diff
@@ -0,0 +1,97 @@
+#include <mcts/SearchLog.hpp>
+
+namespace mcts {
+
+template <core::concepts::Game Game>
+void SearchLog<Game>::build_graph(Graph<Game>& graph) {
+  using State = Game::State;
+  using edge_t = typename Node<Game>::edge_t;
+  auto map = shared_data_->lookup_table.map();
+
+  for (auto [key, node_ix] : *map) {
+    const Node<Game>* node = shared_data_->lookup_table.get_node(node_ix);
+    const State* state = node->stable_data().get_state();
+    graph.add_node(node_ix, node->stats().RN, node->stats().Q,
+                   Game::IO::compact_state_repr(*state));
+    for (int i = 0; i < node->stable_data().num_valid_actions; ++i) {
+      edge_t* edge = node->get_edge(i);
+
+      if (edge->child_index == -1) {
+        continue;
+      }
+
+      int edge_index = i + node->get_first_edge_index();
+      graph.add_edge(edge_index, node_ix, edge->child_index, edge->E, edge->action);
+    }
+  }
+}
+
+template <core::concepts::Game Game>
+inline void SearchLog<Game>::update() {
+  Graph<Game> graph;
+  build_graph(graph);
+  add_graph(graph);
+}
+
+template <core::concepts::Game Game>
+inline boost::json::object SearchLog<Game>::combine_json() {
+  boost::json::array graphs_array;
+  for (auto graph : graphs) {
```

Comment:
Compiler will probably fix this for you with `-O3`, but this is still bad form. You are making a copy of each `Graph` in this for-loop. Fix it by doing:

```
for (const auto& graph : graphs)
```

Reply:
fixed